PluginsAndFeatures/azure-toolkit-for-intellij/src/main/java/com/microsoft/azure/toolkit/intellij/function/runner/core/FunctionCliResolver.java

```diff
@@ -5,73 +5,151 @@
-                    .normalize().toFile();
-            return findFuncExecInFolder(libFolder, true);
+    @Nullable
+    protected String findFuncInNpm() {
+        try {
+            final String output = StringUtils.trim(CommandUtils.exec("npm root --global"));
+            final File path = new File(output, "azure-functions-core-tools/bin");
+            if (FileUtils.isDirectory(path)) {
+                return findFuncInFolder(path.getAbsolutePath());
+            }
+
+        } catch (IOException ignore) {
+            // ignore
         }
+        return null;
     }
 
-    private static File findFuncExecInFolder(final File folder, final boolean windows) {
-        if (new File(folder, "func.dll").exists()) {
-            final File func = new File(folder, windows ? "func.exe" : "func");
-            if (func.exists()) {
-                return func;
+    private static List<String> resolveCommandPath(String command) {
+        final List<String> list = new ArrayList<>();
+        try {
+
+            final String output = CommandUtils.exec((isWindows() ? "where " : "which ") + command);
+            if (StringUtils.isBlank(output)) {
+                return Collections.emptyList();
             }
+
+            for (final String outputLine : output.split("[\\r\\n]")) {
+                final File file = new File(StringUtils.trim(outputLine));
+                if (!file.exists() || !file.isFile()) {
+                    continue;
+                }
+
+                list.add(file.getAbsolutePath());
+            }
+        } catch (IOException ignored) {
+            // ignore
         }
-        return null;
+        return list;
     }
 
+    abstract String getFuncFileName();
+
+    abstract String resolveFuncSystemSpecific(final String funcParentFolder);
+
+    static class WindowsFunctionCoreToolsResolver extends FunctionCliResolver {
+
+        @Override
+        String getFuncFileName() {
+            return "func.exe";
+        }
+
+        @Override
+        String resolveFuncSystemSpecific(String funcParentFolder) {
+            // handle the case of choco install:
+            // from C:\ProgramData\chocolatey\bin\func.exe -> C:\ProgramData\chocolatey\lib\azure-functions-core-tools\tools\func.exe
+            return findFuncInFolder(Paths.get(funcParentFolder, "..\\lib\\azure-functions-core-tools\\tools").toString());
+        }
+    }
+
+    static class LinuxFunctionCoreToolsResolver extends FunctionCliResolver {
+        @Override
+        String getFuncFileName() {
+            return "func";
+        }
+
+        @Override
+        String resolveFuncSystemSpecific(final String funcParentFolder) {
+            // detect func installed by `brew install azure-functions-core-tools@3`
+            // readlink /usr/local/bin/func =>
+            //../Cellar/azure-functions-core-tools@3/3.0.3477/bin/func
+            return findFuncInFolder(Paths.get(funcParentFolder, "../bin").toString());
```

Comment:
only relative folder are different ?
how about to define a  abstract "getRelativeFuncFolder" method in their parent to replace `resolveFuncSystemSpecific` with ?

Reply:
for func, I will update it in later PR